### PR TITLE
fast: optional fused residual-add in rms_norm (Metal)

### PR DESCRIPTION
**Pull request by jonathan308**
Every decoder layer runs `h = x + sublayer_out; normed = rms_norm(h)` twice — two kernel launches and an extra full-width read/write of the stream per boundary (72 boundaries per token at 36 layers). This PR adds an optional `residual` argument to `mx.fast.rms_norm`:

```python
normed, h = mx.fast.rms_norm(x, weight, eps, residual=r)
# normed == rms_norm(x + r) * weight ; h == x + r
```

- `residual=None` (default) → unchanged behavior, single array out. Fully backward compatible.
- Metal: two fused kernels (register-cached single-row for H≤3072, looped for larger H) reusing the existing threadgroup reduction structure.
- VJP implemented (`RMSNormVJP` on the sum + add for the sum cotangent); verified against composed autograd (<1e-5 rel).
- CUDA/no-GPU: composed fallback (correct, unfused); a fused CUDA kernel is a straightforward follow-up.

### Numerics
`out_sum` bit-identical to `x + r`. `out` bit-identical for H>3072; ≤1 ulp (fp16/bf16) for H≤3072 (reduction regrouping; documented in tests). Model-level: Qwen3-8B logits **bit-identical** fused vs unfused.

### Benchmarks (M3 Ultra)
- Micro, GPU-bound shapes (4,512,H): 1.13–1.38x for H≥5120, all dtypes. Decode shapes (1,1)/(1,8) are submission-bound: parity to 1.14x.
- E2E decode (mlx_lm, Qwen3-8B-4bit with call sites adopted, interleaved A/B): this PR contributes ~+1.1–2.6% decode tok/s on top of the numbers below, more at longer context.
- Honest losses in micro at small-H batched cells (harness cross-iteration overlap artifacts; convert to wins in serialized dependent chains) are documented in the test comments.

### Tests
- `test_rms_norm_residual` / `test_rms_norm_residual_grad` added (54 dtype/shape configs); full `test_fast.py` 29/29 pass (standalone build, this branch).

### Companion adoption
An mlx-lm PR adopting `residual=` at decoder-layer boundaries (feature-detected so older mlx keeps working) will follow; it is deliberately held until this API lands in a release.

### CODE REVIEW

**Comment by zcbenz** [comment]
Thanks for the PR but at the moment we do not plan to add new fast ops unless it is battle tested, see also https://github.com/ml-explore/mlx/issues/4259#issuecomment-5301003296.
